### PR TITLE
goToTypeDefinition: Go to function return type

### LINKED ISSUE #25933: Go-to-type is not useful for function
**TypeScript Version:**  3.1.0-dev.20180725

**Search Terms:**

**Code**

```ts
interface I { x: number; }
function f(): I { return { x: 0 }; }

f();
```

Go-to-type on `f()`.

**Expected behavior:**

Takes me to `I`.

**Actual behavior:**

Takes me to `f` -- same as go-to-definition.

Contrast with this Kotlin code:

```kt
class C(val x: Int)

fun f() = C(0)

fun main() {
    f()
}
```

where go-to-declaration on `f()` takes me to the definition, but go-to-type takes me to `C`.

### PR DESCRIPTION
**Pull request by andy-ms**
Fixes #25933

### CODE REVIEW

File: tests/cases/fourslash/server/declarationMapsEnableMapping_NoInline.ts [hunk 1]
```diff
@@ -76,7 +76,7 @@ goTo.file("/index.ts");
 verify.getEmitOutput(["/dist/index.js", "/dist/index.d.ts.map", "/dist/index.d.ts"]);
 
 verify.goToDefinition("1", "2"); // getDefinitionAndBoundSpan
-verify.goToType("1", "2"); // getTypeDefinitionAtPosition
+verify.goToType("1", "SomeType"); // getTypeDefinitionAtPosition
```
Comment (sheetalkamat) [inline tests/cases/fourslash/server/declarationMapsEnableMapping_NoInline.ts]: can you add test case where method returns `void`

File: tests/cases/fourslash/goToTypeDefinition_returnType.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+////const f2 = (i: I): I => ({ x: i.x + 1 });
+////
+/////*f0*/f0();
+/////*f1*/f1();
+/////*f2*/f2();
+
+verify.goToType({
+    f0: "I",
+    f1: "T",
+    f2: "I",
```
Comment (sheetalkamat) [inline tests/cases/fourslash/goToTypeDefinition_returnType.ts]: Can you add test case where no return type is specified and it is inferred

Comment (sheetalkamat) [reply tests/cases/fourslash/goToTypeDefinition_returnType.ts]: Another one when 
- return type is union
- inferred type is union

And while at it also test case with overload be good to have.

**Comment by andy-ms** [comment]
@sheetalkamat Please re-review

File: tests/cases/fourslash/goToTypeDefinition_returnType.ts [hunk 3]
```diff
@@ -9,12 +10,38 @@
...
 
 verify.goToType({
     f0: "I",
     f1: "T",
     f2: "I",
+    f3: "f3Def",
+    f4: "I",
+    f5: ["I", "J"],
+    f6: ["I", "J"],
+    f7: [],
```
Comment (sheetalkamat) [inline tests/cases/fourslash/goToTypeDefinition_returnType.ts]: Shouldnt this go to f7Def ?

**Pull request merged**